### ISSUE TITLE
Functional API and Plain JS Objects

### ISSUE THREAD

**Issue by dselman** [accordproject/concerto#187]
Concerto would benefit from a non class based API - allowing Concerto objects to be plain old JS objects. You would no longer need to use the `Factory` to create them, or the `Serializer` to convert them to/from plain JS objects.

**Describe the solution you'd like**
Examine all the methods on Resource, Identifiable, Typed etc and move them into a set of functions. The functions could be static methods on a class, or just exported as a library of functions.

E.g.

```
const modelManager = new ModelManager()
...
const myObject = { ... }
Concerto.validate( myObject, modelManager );
```

**Describe alternatives you've considered**

Messing with the prototype: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object

**Additional context**

Removing the need to use the Factory and Serializer to create Concerto objects would simplify the APIs, learning curve and could lead to performance improvements.

> This new API could be developed in parallel to the existing API, allowing for a progressive migration strategy.

Note that we don't necessarily need to change the core internal APIs (things like `ModelManager`).

Here are the instance-based APIs which would need to be re-implemented as functions.

```
class Concept extends Typed {
   + boolean isConcept() 
}
class Identifiable extends Typed {
   + string getIdentifier() 
   + void setIdentifier(string) 
   + string getFullyQualifiedIdentifier() 
   + String toString() 
   + boolean isRelationship() 
   + boolean isResource() 
   + String toURI() 
}
class Relationship extends Identifiable {
   + String toString() 
   + boolean isRelationship() 
   + Relationship fromURI(ModelManager,String,String,String) 
}
class Resource extends Identifiable {
   + String toString() 
   + boolean isResource() 
   + Object toJSON() 
}
class Typed {
   + string getType() 
   + string getFullyQualifiedType() 
   + string getNamespace() 
   + void setPropertyValue(string,string) 
   + void addArrayValue(string,string) 
   + boolean instanceOf(String) 
}
class ValidatedConcept extends Concept {
   + void setPropertyValue(string,string) throws Error
   + void addArrayValue(string,string) throws Error
   + void validate() throws Error
}
class ValidatedResource extends Resource {
   + void setPropertyValue(string,string) throws Error
   + void addArrayValue(string,string) throws Error
   + void validate() throws Error
}
```

**Comment by jeromesimeon** [comment]
I love this idea. Some fairly high-level questions.

- Do we still need all / any of the classes in `model` or would we work directly against the JSON data? I think the later would be great.
- What happens when we remove one of those methods right now, where does it break.
- The visitor pattern is somewhat intimately bound to the structure of the internal classes at the moment. Should we also review/revise the way we do those transforms.
- I think with this approach from JSON could be replaced by validation. to JSON could disappear entirely.

**Comment by jeromesimeon** [comment]
A bit of a work list, wondering if this makes sense @dselman 
-[ ] Remove the classes in `./model` (maybe except resources -- not entirely clear on this)
-[ ] Remove or replace the ./factory` class which is used quite a bit through the stack. It already has a mostly functional interface so may be something to keep, but based on the new API?
-[ ] Remove or replace the to/from serializers. This one has a big impact, including on: markdown-transform, ergo, Cicero execution
-[ ] Decide whether we need any alternative transformation later (keep the visitors which are model-driven? add a data-driven functional transformer?) -- this may not be as bad as I thought and could happen later since currently we mostly use the

-[ ] Review impact through the stack, the markdown transform uses the factory and direct manipulation of the current object classes quite a bit. Direct JSON access would be I believe a big improvement, but may be a non-trivial amount of work.

**Comment by dselman** [comment]
> I love this idea. Some fairly high-level questions.
> 
> * Do we still need all / any of the classes in `model` or would we work directly against the JSON data? I think the later would be great.

These can be removed for 1.0.

> * What happens when we remove one of those methods right now, where does it break.

It tends to break in the visitors, which usually assume that they are dealing with a Concerto object. The visitors will have to be updated to use the new `Concerto` API to get the type name, namespace etc. This is work, but not terribly complex to refactor.

> * The visitor pattern is somewhat intimately bound to the structure of the internal classes at the moment. Should we also review/revise the way we do those transforms.

I suggest that leave visiting the model as-is for now - as that is a much bigger change.

> * I think with this approach from JSON could be replaced by validation. to JSON could disappear entirely.

Yes - please see the new `Concerto` API and tests for examples.

**Comment by dselman** [comment]
> A bit of a work list, wondering if this makes sense @dselman
> 
> * [ ]  Remove the classes in `./model` (maybe except resources -- not entirely clear on this)

For 1.0 only.

> * [ ]  Remove or replace the ./factory` class which is used quite a bit through the stack. It already has a mostly functional interface so may be something to keep, but based on the new API?

Remove for 1.0.

> * [ ]  Remove or replace the to/from serializers. This one has a big impact, including on: markdown-transform, ergo, Cicero execution

Remove for 1.0.

> * [ ]  Decide whether we need any alternative transformation later (keep the visitors which are model-driven? add a data-driven functional transformer?) -- this may not be as bad as I thought and could happen later since currently we mostly use the model-driven visitor.

Agree. Let's do this later.

> * [ ]  Review impact through the stack, the markdown transform uses the factory and direct manipulation of the current object classes quite a bit. Direct JSON access would be I believe a big improvement, but may be a non-trivial amount of work.

Agree - but the refactoring should be fairly "mechanical" and not too complex, based on the new `Concerto` API.

**Comment by dselman** [comment]
Code now has a set of "positive" set of tests which pass. There are still some negative tests to add: data which fails to validate to the schema.

Open issues:

1. `DataTime` fields use `moment-mini` to validate that the string parses as an ISO-8601 date.

2. No handling for default values in the model. This code currently lives in `Typed.assignFieldDefaults`. I think we will need a specific function for this:

```
const in = { ... };
const out = assignDefaults( in, modelManager );
```

3. We need to remove the Ergo specific code from Concerto

**Comment by jeromesimeon** [comment]
> 1. We need to remove the Ergo specific code from Concerto

This is issue https://github.com/accordproject/concerto/issues/162

**Comment by dselman** [comment]
Closing this, as the new Concerto class has been merged into the release-1.0 branch.
